### PR TITLE
[Truffle] Added a new internal method for taking ownership of an array store from another array.

### PR DESCRIPTION
**Pull request by nirvdrum**
This is intended to be a faster version of Array#replace when we know that the source array won't be used anywhere else.

While escape analysis may very well catch some of these, I don't think it would catch all of them (e.g., the mergesort! case). And this will work in the interpreter as well.

@jruby/truffle

### CODE REVIEW

**Comment by chrisseaton** [comment]
Can you write a quick couple of specs for this in `spec/truffle/specs/truffle/array`.

**Comment by pitr-ch** [comment]
👍

**Comment by eregon** [comment]
Just a nitpick on the name: `store` is also the name of the field for the backing Java array. Maybe call this `storage`?
For the full name I like `steal_storage` and also `become` as suggested by Chris.
`swap` might be an interesting way to model this too, with the added semantics of exchanging the storages.

**Comment by nirvdrum** [comment]
Fair enough. I named it with "store" to make it clear that you're updating that field. This is an internal operation.

File: truffle/src/main/java/org/jruby/truffle/core/array/TruffleArrayNodes.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import static org.jruby.truffle.core.array.ArrayHelpers.getSize;
+import static org.jruby.truffle.core.array.ArrayHelpers.getStore;
+import static org.jruby.truffle.core.array.ArrayHelpers.setStoreAndSize;
+
+@CoreClass("Truffle::Array")
+public class TruffleArrayNodes {
+
+    @CoreMethod(names = "steal_storage", onSingleton = true, required = 2)
+    public abstract static class TakeOwnershipOfStoreNode extends CoreMethodArrayArgumentsNode {
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/core/array/TruffleArrayNodes.java]: (the name of the node should also be updated)

Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/core/array/TruffleArrayNodes.java]: Thanks. I obviously missed that.

**Pull request merged**